File: hotel-mern/services/room.service.ts
import { RoomPopulated } from '../shared/models'
import { RoomRepository } from '../shared/repositories'
import { IQuery } from '../shared/interfaces'

export default class RoomService extends RoomRepository {
  async get(
    query: IQuery,
    limit: number = Number.MAX_SAFE_INTEGER,
    offset: number = 0
  ): Promise<RoomPopulated[]> {
    return this.model
      .find(query)
      .skip(offset)
      .limit(limit)
      .populate('_type')
      .populate('_building')
      .lean()
  }

  async getOne(_id: string): Promise<RoomPopulated> {
    return this.model
      .findById(_id)
      .populate('_type')
      .populate('_building')
      .lean()
  }

  async getAmount(query: IQuery): Promise<number> {
    return this.model.find(query).countDocuments()
  }

  async create(_building: string, _type: string): Promise<RoomPopulated> {
    const room = await new this.model({
      _building,
      _type,
      isFree: true,
      population: 0,
    })
    await room.save()
    return this.getOne(room._id)
  }

  async change(
    _id: string,
    _building: string,
    _type: string
  ): Promise<RoomPopulated> {
    await this.model
      .findByIdAndUpdate(_id, {
        $set: { _building, _type },
      })
      .lean()
    return this.getOne(_id)
  }

  async bookRoom(roomId: string, orderId: string, population: number) {
    await this.model.findByIdAndUpdate(roomId, {
      $set: { _order: orderId, isFree: false, population },
    })
  }

  async unBookRoom(roomId: string) {
    await this.model.findByIdAndUpdate(roomId, {
      $set: { isFree: true, population: 0 },
    })
    await this.model.findByIdAndUpdate(roomId, { $unset: { _order: '' } })
  }

  async delete(_id: string): Promise<string> {
    await this.model.findByIdAndRemove(_id)
    return _id
  }

  async deleteWithBuilding(_building: string) {
    await this.model.deleteMany({ _building })
  }

  async deleteWithType(_type: string) {
    await this.model.deleteMany({ _type })
  }
}

File: hotel/backend/src/main/kotlin/loshica/api/hotel/services/RoomService.kt
package loshica.api.hotel.services

import loshica.api.hotel.core.BaseService
import loshica.api.hotel.interfaces.IRoomService
import loshica.api.hotel.models.*
import loshica.api.hotel.repositories.RoomRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class RoomService(
    @Autowired override val repository: RoomRepository
) : BaseService<Room, RoomRepository>(repository), IRoomService {

    private fun getByQuery(
        buildingId: Int?,
        typeId: Int?,
        isFree: Boolean?
    ): Iterable<Room> {
        var rooms: Iterable<Room> = repository.findAll()

        buildingId?.let {
            rooms = rooms.filter { room -> room.building.id == it }
        }

        typeId?.let {
            rooms = rooms.filter { room -> room.type.id == it }
        }

        isFree?.let {
            rooms = rooms.filter { room -> room.isFree == it }
        }

        return rooms
    }

    override fun get(
        buildingId: Int?,
        typeId: Int?,
        isFree: Boolean?,
        limit: Int,
        offset: Int
    ): Iterable<Room> {
        val rooms: Iterable<Room> = getByQuery(buildingId, typeId, isFree)

        return rooms
            .withIndex()
            .filter { indexedValue: IndexedValue<Room> ->
                indexedValue.index >= offset && indexedValue.index <= offset + limit
            }
            .map { indexedValue: IndexedValue<Room> -> indexedValue.value }
    }

    override fun getAmount(buildingId: Int?, typeId: Int?, isFree: Boolean?): Int {
        return getByQuery(buildingId, typeId, isFree).count()
    }

    override fun create(building: Building, type: Type): Room {
        val room = Room(building = building, type = type)
        repository.save(room)
        return room
    }

    override fun change(id: Int, building: Building, type: Type): Room {
        val room: Room = getOne(id)
        room.change(building = building, type = type)
        repository.save(room)
        return room
    }

    override fun bookRoom(booking: Booking) {
        booking.room.let {
            it.book(booking)
            repository.save(it)
        }
    }

    override fun unBookRoom(booking: Booking) {
        booking.room.let {
            it.unBook(booking)
            repository.save(it)
        }
    }

    override fun addComment(comment: Comment) {
        comment.room.let {
            it.comments.add(comment)
            repository.save(it)
        }
    }

    override fun removeComment(comment: Comment) {
        comment.room.let {
            it.comments.remove(comment)
            repository.save(it)
        }
    }
}